Docs/reformat service splitters conf entry (#16264)
hashicorp/consul: website/content/docs/connect/config-entries/service-splitter.mdx

--- 
Page_title: Service Splitter | Configuration Reference

> Description: 

Reviewer: Should this be empty?

> Description: 

Author: ```suggestion
Description: Service splitter config entries are L7 traffic management tools that enable Consul to redirect parts of incoming traffic requests for a specific service. 
```

---

# Service Splitter Configuration Reference

> This topic describes how to configure service splitter configuration entries. Configure and apply service splitters to redirect a percentage of incoming traffic requests for a specific service.

Author: 1. I like "topic" as a general way to talk about a unit of information, but I don't have a problem with "reference page". 
2. My thinking is that it might be clearer to refer to the "service splitter configuration entry" when talking about the entity in general and use `service-splitter` and `ServiceSplitter` when referring to the VM and YAML configurations respectively.

	  - [`serviceSubset`](#spec-splits-servicesubset): string
    - [`namespace`](#spec-splits-namespace): string <EnterpriseAlert inline />
    - [`partition`](#spec-splits-partition): string <EnterpriseAlert inline />
    - [`requestHeaders`](#spec-splits-requestheaders): HTTPHeaderModifiers 
      - [`add`](#spec-splits-requestheaders): map 
      - [`set`](#spec-splits-requestheaders): map 
      - [`remove`](#spec-splits-requestheaders): map 
    - [`responseHeaders`](#spec-splits-responseheaders): HTTPHeaderModifiers 
      - [`add`](#spec-splits-responseheaders): map 
      - [`set`](#spec-splits-responseheaders): map 
      - [`remove`](#spec-splits-responseheaders): map 

</Tab>
</Tabs>

## Complete configuration

When every field is defined, a service splitter configuration entry has the following form:

<Tabs>

<Tab heading="HCL" group="hcl">

```hcl
Kind      = "service-splitter"           ## string  | required
Name      = "config-entry-name"          ## string  | required
Namespace = "main"                       ## string
Partition = "partition"                  ## string 
Meta      = {                            ## map
  key     = "value"            

>   }

Reviewer: The formatting on these indented objects looks a little strange. I think it's because the `=` symbols are lined up. How do we usually format HCL/JSON?

I'd expect it to look more like `gofmt`'s formatting of struct fields:

```Go
foo := Foo{
	AAA:  "a",
	BBBB: "b",
	CCCCC: Other{
		D: "d",
	},
}
```

Note the scalar field values are lined up with each other, but not `CCCCC` or its fields.

Author: @boxofrad Just to be clear, you're saying that we should line  values up, not the symbols?, so:

```
Kind =      "service-splitter"
Name =      "config-entry-name"
Namespace = "main"                   
Partition = "partition"  
```

Reviewer: Sorry, my bad for not being clear! Go with whatever the established style is, but I'd expect the above snippet to look more like this:

```hcl
Kind      = "service-splitter"
Name      = "config-entry-name"
Namespace = "main"
Partition = "partition"
Meta = {
  key       = "value"
  other_key = "value"            
}
```

Actually, maybe run [`hclfmt`](https://github.com/fatih/hclfmt) on it?

kind: ServiceSplitter                             # string | required
metadata:                                         # object | required
  name: config-entry-name                         # string | required
  namespace: main                                 # string
spec:
  splits:                                         # list
    - weight: 90                                  # floating point | required
      service: service                            # string 
      serviceSubset: v1                           # string 
      namespace: target-namespace                 # string 
      partition: target-partition                 # string 
      requestHeaders:  
        set:
          x-web-version: from-v1                  # string 
      responseHeaders:
        set:
          x-web-version: to-v1                    # string 
    - weight: 10
      service: service
      serviceSubset: v2
      namespace: target-namespace
      partition: target-partition
      requestHeaders:
        set:
          x-web-version: from-v2
      responseHeaders:
        set:
          x-web-version: to-v2      
```

</Tab>

</Tabs>

## Specification

> This section provides details about the fields you can configure in the sservice splitter configuration entry.

Reviewer: ```suggestion
This section provides details about the fields you can configure in the service splitter configuration entry.
```

<Tabs>

<Tab heading="HCL" group="hcl">

### `Kind` 

Specifies the type of configuration entry to implement. 

#### Values

- Default: none
- This field is required
- Data type: String value that must be set to `service-splitter` in HCL or `serviceSplitter` in YAML. 

### `Name`

Specifies a name for the configuration entry. The name is metadata that you can use to reference the configuration entry when performing operations, such as reading or writing configuration entry configuration entry settings. 

#### Values

> - Default: Name of the node 

Author: Yeah, kind of wordy. I think "default" means the same thing as "If left empty, this value is applied". I tried something half way.

- This field is required
- Data type: String

### `Namespace` <EnterpriseAlert inline />

Specifies the [namespace](/consul/docs/enterprise/namespaces) to apply the configuration entry.

#### Values

- Default: None

> - Data type: String

Author: Yeah, let's add this to the topics to discuss. My preference is probably plaintext over code font because the user isn't actually writing the word "string" 🤷 , but we should still use a consistent case.

### `Partition` <EnterpriseAlert inline />
Specifies the [admin partition](consul/docs/enterprise/admin-partitions) to apply the configuration entry.

#### Values

> - Default: `Default`   

Author: I don't think that the default admin partition is capitalized

- Data type: String

### `Meta`

Specifies key-value pairs to add to the KV store.

#### Values

- Default: none 
- Data type: Map of one or more key-value pairs 
  - keys: String
  - values: String, integer, or float 

### `Splits`

Defines how much traffic to send to sets of service instances during a traffic split. 

#### Values 

- Default: None
- This field is required for defining service splits
- Data type: list of objects that can contain the following fields:
  - `Weight`: The sum of weights for a set of service instances must add up to 100. 

>   - `Service`: This field is required to direct split traffic   

Author: My thought is that technically, the conf entry will work with or without all fields except for kind and name, I believe. At the end of the day, yeah, the reason you're doing this is to split traffic, so maybe it's _splitting_ hairs.

- Data type: String

### `spec.splits[].serviceSubset`

Specifies a subset of the service to resolve. This field overrides the `DefaultSubset`. 

#### Values

- Default: If empty, the `split` uses the default subset. 
- Data type: String

### `spec.splits[].namespace` <EnterpriseAlert inline />

Specifies which [namespace](/consul/docs/enterprise/namespaces) to use for resolving the service. 

#### Values

- Default: The namespace specified in the top-level configuration entry.
- Data type: String

### `spec.splits[].partition` <EnterpriseAlert inline />

Specifies which [admin partition](/consul/docs/enterprise/admin-partitions) to use in the FQDN when resolving the service. 

#### Values

- Default: By default, the `service-splitter` uses the [admin partition specified in the top-level configuration entry](#spec-partition).
- Data type: String

### `spec.splits[].requestHeaders`

Specifies a set of HTTP-specific header modification rules that apply to requests routed to the split. You cannot configure request headers if the listener protocol is set to `tcp`. Refer to [Set HTTP Headers](#set-http-headers) for an example configuration. 

>  #### Values

Reviewer: I think the trailing space will prevent this from being rendered as a header.
```suggestion
#### Values
```

| `set` | Defines a set of key-value pairs to add to the request header or to replace existing header values with. Use header names as the keys. Header names are not case-sensitive. If header values with the same names already exist, Consul replaces the header values. You can [use variable placeholders](#use-variable-placeholders). | map of strings | 
| `remove` | Defines an list of headers to remove. Consul removes only headers containing exact matches. Header names are not case-sensitive. | list of strings | 

#### Use variable placeholders

For `add` and `set`, if the service is configured to use Envoy as the proxy, the value may contain variable placeholders, for example, `%DOWNSTREAM_REMOTE_ADDRESS%` to interpolate dynamic metadata into the value added.

### `spec.splits[].responseHeaders`

Specifies a set of HTTP-specific header modification rules that apply to responses from the split. You cannot configure request headers if the listener protocol is set to `tcp`. Refer to [Set HTTP Headers](#set-http-headers) for an example configuration.

>  #### Values

Reviewer: ```suggestion
#### Values
```

### `Splits`

Defines how much traffic to send to sets of service instances during a traffic split. 

#### Values 

- Default: None
- This field is required for defining service splits
- Data type: list of objects that can contain the following fields:
  - `Weight`: The sum of weights for a set of service instances must add up to 100. 
  - `Service`: This field is required to direct split traffic   
  - `ServiceSubset`
  - `Namespace` <Enterprise inline/>
  - `Partition` <Enterprise inline/>
  - `RequestHeaders`
  - `ResponseHeaders`  

### `Splits[].Weight`

Specifies the percentage of traffic sent to the set of service instances specified in the [`Service`](#service) field. Each weight must be a floating integer between `0` and `100`. The smallest representable value is `.01`. The sum of weights across all splits must add up to `100`.

#### Values

- Default: `null`
- This field is required for defining service splits
- Data type: Floating integer from `.01` to `100`

### `Splits[].Service`

Specifies the name of the service to resolve.

#### Values

- Default: The service matching the configuration entry [`Name`](#name) field.
- Data type: String

### `Splits[].ServiceSubset`

> Specifies a subset of the service to resolve. This field overrides the `DefaultSubset`. 

Author: ```suggestion
Specifies a subset of the service to resolve. A service subset assigns a name to a specific subset of discoverable service instances within a datacenter, such as `version2` or `canary`. All services have an unnamed default subset that returns all healthy instances. 

You can define service subsets in a [service resolver configuration entry](/consul/docs/connect/config-entries/service-resolver), which are referenced by their names throughout the other configuration entries. This field overrides the default subset value in the service resolver configuration entry.   
```
Thanks for catching this - I was stumped for a while and had forgotten to come back to it.
